app/models/job_run.rb

```diff
@@ -0,0 +1,44 @@
+class JobRun < ApplicationRecord
+  validates :host, presence: true
+  validates :pid, presence: true
+
+  after_initialize :set_default_values
+
+  def self.with_lock
+    raise ArgumentError, 'Must pass block' unless block_given?
+    transaction do
+      connection.execute(
+        "LOCK #{connection.quote_table_name(table_name)} IN ACCESS EXCLUSIVE MODE",
+      )
+      # Yield to caller block with the lock held
+      yield
+    end
+  end
+
+  def self.clean_up_timeouts(job_name:, timeout_threshold:)
+    # Find all runs that did not finish and don't have an error recorded that
+    # are older than the timeout threshold.
+    where(job_name: job_name).where(finish_time: nil).where(error: nil).
+      where('created_at < ?', timeout_threshold).
+      find_each(&:mark_as_timed_out)
```

Comment:
I just realized this isn't safe to run concurrently, or else we may receive many timeout notices for the same job.

I believe we can use the built-in support for row-level locking for this case:

https://api.rubyonrails.org/classes/ActiveRecord/Locking/Pessimistic.html

It looks like you can just chain `.lock` after the `.where()` and it will do the right thing.

Or we could reuse the existing `.with_lock` from above and take out a whole table lock, although that's less granular.